includes/subscriptions/class-wc-payments-invoice-service.php

```diff
@@ -110,136 +166,79 @@ public function maybe_record_first_invoice_payment( int $order_id, string $old_s
-
-		if ( $discount ) {
-			$discount_data             = $this->format_invoice_item_data( -$discount, $currency, __( 'Discount', 'woocommerce-payments' ) );
-			$discount_data['customer'] = $wcpay_customer_id;
-
-			if ( ! empty( $wcpay_subscription_id ) ) {
-				$discount_data['subscription'] = $wcpay_subscription_id;
-			}
-
-			$data[] = $discount_data;
-		}
-
-		$items = $is_new ? $subscription->get_items() : [];
+		foreach ( $wcpay_items as $item ) {
+			$wcpay_subscription_item_id = $item['subscription_item'];
 
-		if ( ! $is_delayed ) {
-			$shipping_items = $is_new ? $subscription->get_parent()->get_shipping_methods() : $subscription->get_shipping_methods();
-			$items          = array_merge( $items, $subscription->get_fees(), $shipping_items );
+			$wcpay_item_data[ $wcpay_subscription_item_id ] = [
+				'amount'    => $item['amount'],
+				'quantity'  => $item['quantity'],
+				'tax_rates' => array_column( $item['tax_rates'], 'percentage' ),
+			];
 		}
 
-		foreach ( $items as $item ) {
-			$is_line_item = $item->is_type( 'line_item' );
-			$amount       = $is_line_item ? floatval( WC_Subscriptions_Product::get_sign_up_fee( $item->get_product() ) ) : $item->get_total();
+		foreach ( $subscription->get_items( [ 'line_item', 'fee', 'shipping' ] ) as $item ) {
+			$subscription_item_id = WC_Payments_Subscription_Service::get_wcpay_subscription_item_id( $item );
 
-			if ( $amount ) {
-				$description = $is_line_item ? __( 'Sign-up Fee', 'woocommerce-payments' ) : ucfirst( $item->get_name() );
-				$tax_rates   = $this->product_service->get_tax_rates_for_item( $item, $subscription );
+			if ( ! $subscription_item_id ) {
+				continue;
+			}
 
-				$item_data             = $this->format_invoice_item_data( $amount, $currency, $description, $tax_rates );
-				$item_data['customer'] = $wcpay_customer_id;
+			if ( ! in_array( $subscription_item_id, array_keys( $wcpay_item_data ), true ) ) {
+				$message = __( 'The WCPay invoice items do not match WC subscription items', 'woocommerce-payments' );
+				Logger::error( $message );
+				throw new Rest_Request_Exception( $message );
+			}
 
-				if ( ! empty( $wcpay_subscription_id ) ) {
-					$item_data['subscription'] = $wcpay_subscription_id;
+			$item_data   = $wcpay_item_data[ $subscription_item_id ];
+			$repair_data = [];
+
+			if ( (int) $item->get_total() * 100 !== $item_data['amount'] ) {
+				if ( $item->is_type( 'line_item' ) ) {
+					$product              = $item->get_product();
+					$repair_data['price'] = $this->product_service->get_stripe_price_id( $product );
+				} else {
+					$repair_data['price_data'] = WC_Payments_Subscription_Service::format_item_price_data(
+						$subscription->get_currency(),
+						$this->product_service->get_stripe_product_id_for_item( $item->get_type() ),
+						$item->get_total(),
+						$subscription->get_billing_period(),
+						$subscription->get_billing_interval()
+					);
```

Comment:
Re this comment: https://github.com/Automattic/woocommerce-payments/pull/2899#issuecomment-922695478

I think we need to add the items tax data similar to how we do it here: https://github.com/Automattic/woocommerce-payments/blob/update/subscription_items/includes/subscriptions/class-wc-payments-subscription-service.php#L637

Reply:
We are actually adding tax data further down in this method. Since this is updating subscription items rather than creating them, we only include the repair data that is needed with the update request.

As for this comment: https://github.com/Automattic/woocommerce-payments/pull/2899#issuecomment-922695478

Good catch again! I believe I've fixed this here: https://github.com/Automattic/woocommerce-payments/pull/2899/commits/afda4e237b2a0408de285c70dd846012badd44b4

Basically, before this change, we were validating tax data by checking that each of our tax rates in WC matched a tax rate in Stripe. But we weren't accounting for cases where Stripe had somehow added an extra tax rate in addition to any that we already were expecting. This change first checks that Stripe and WC have the same number of tax rates per item before iterating through each WC tax rate to compare actual percentages.

Reply:
> But we weren't accounting for cases where Stripe had somehow added an extra tax rate in addition to any that we already were expecting.

oooh. Good thing you picked that detail out of my rough steps to replicate. 😅